Add support for a second Ethernet controller in the MCUX driver

armandciejak opened the pull request:
Closes #23284.

This PR allows using the second Ethernet controller of the i.MXRT1060 family.

armandciejak committed:
drivers: eth: eth_mcux: Add 2nd Ethernet controller to RT1060 device tree

Add eth1 in dtsi and defines for the second controller in fixup.

Signed-off-by: Armand Ciejak <armandciejak@users.noreply.github.com>

armandciejak committed:
drivers: eth: mcux: Add ETH_MCUX_1 config

Add new config allowing to enable the second Ethernet port
on i.MX RT1060 SoC family.

Signed-off-by: Armand Ciejak <armandciejak@users.noreply.github.com>

armandciejak committed:
drivers: eth: mcux: Add ENET base address to eth_context structure

This will allow using either ENET or ENET2.

Signed-off-by: Armand Ciejak <armandciejak@users.noreply.github.com>

armandciejak committed:
drivers: eth: mcux: Add config_func to eth_context structure

This will allow having a different function for the second
port.

Signed-off-by: Armand Ciejak <armandciejak@users.noreply.github.com>

armandciejak committed:
drivers: eth: mcux: Add generate_mac function pointer to eth_context

This will allow for different configuration between ENET and ENET2.

Signed-off-by: Armand Ciejak <armandciejak@users.noreply.github.com>

armandciejak committed:
drivers: eth: mcux: Add PHY address to eth_context

This will allow having a different PHY address for each controller.
It also avoids for the PHY address to be hard-coded in several
functions.

Signed-off-by: Armand Ciejak <armandciejak@users.noreply.github.com>

armandciejak committed:
drivers: eth: mcux: Add eth1 for i.MXRT1060 family

Instantiate a second Ethernet device if CONFIG_ETH_MCUX_1 is set.

Signed-off-by: Armand Ciejak <armandciejak@users.noreply.github.com>

armandciejak commented:
It is mostly ready, but I still need to do more testing before removing WIP status. Please review and comment.

I also plan to make the PHY address defined in the device tree, but I'm still undecided if I do it in a separated PR or not.

armandciejak commented:
I've successfully tested my changes on the MIMXRT1060-EVK with the following test apps:

- samples/net/dhcpv4_client
- samples/net/sockets/dumb_http_server
- samples/net/sockets/big_http_download

agansari commented on dts/arm/nxp/nxp_rt1060.dtsi:
```diff
@@ -16,3 +16,29 @@
+		eth1 = &eth1;
+	};
+
+	soc {
```
please `/delete-node/ &eth1;` in _boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts_ and _mimxrt1064_evk_ 
SoC has 2 interfaces, but these boards don't have the second inteface wired.

armandciejak replied:
Done in 46e5bdddb4ecf414eb8c3c3d734a623741a11483

armandciejak committed:
drivers: eth: mcux: Delete eth1 node for EVK boards

These boards have a SoC with 2 Ethernet controllers but the second
controller it is not wired.

Signed-off-by: Armand Ciejak <armandciejak@users.noreply.github.com>

agansari commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -897,7 +912,7 @@ static int eth_0_init(struct device *dev)
 	eth_mcux_phy_setup(context);
 
-	LOG_DBG("MAC %02x:%02x:%02x:%02x:%02x:%02x",
+	LOG_DBG("[%p] MAC %02x:%02x:%02x:%02x:%02x:%02x", context->base,
```
I would prefer printing DT_ETH_MCUX_x_NAME or instance number rather than the address in context->base.

armandciejak replied:
I'll do that.

armandciejak replied:
I changed it in 99237a1472faf2daf552a768ce18daf904a06f3f

agansari reviewed (changes requested):
Mostly looks okay, I need to test this in FRDM-K64F (single interface Kinetis).

armandciejak commented on dts/arm/nxp/nxp_rt1060.dtsi:
```diff
@@ -16,3 +16,29 @@
+
+/* i.MX rt1060 has a second Ethernet controller.
+ */
+/ {
```
OK for the comment, but I don't see why putting the { to a separate line.

armandciejak replied:
Do you mean like this?

```
/* i.MX rt1060 has a second Ethernet controller. */

/ {
```

With a blank line between the comment and `/ {`

agansari replied:
Looks good

armandciejak commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -284,9 +285,9 @@ static void eth_mcux_phy_event(struct eth_context *context)
 	case eth_mcux_phy_state_initial:
 #ifdef CONFIG_SOC_SERIES_IMX_RT
-		ENET_StartSMIRead(ENET, phy_addr, PHY_CONTROL2_REG,
+		ENET_StartSMIRead(context->base, phy_addr, PHY_CONTROL2_REG,
```
Since context->base is not volatile the compiler will optimize the dereferencing of the value when necessary. Or do you suggest to use a local variable for readability?

agansari replied:
mostly readability, i assume the compiler is smart enough to optimize.

armandciejak replied:
I'm not convinced about the benefit for readability, but if you insist I can do it and if I do it for `context->base` I think it make sense to do it for `context->phy_addr` as well.

armandciejak replied:
@agansari Please tell me If you want me to do the changes or not.

agansari replied:
@armandciejak i'm okay with your current implementation this is why I gave it a +1

armandciejak commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -208,7 +208,8 @@ static void eth_mcux_phy_enter_reset(struct eth_context *context)
 {
 #ifdef CONFIG_ETH_MCUX_PHY_EXTRA_DEBUG
-	LOG_DBG("phy_state=%s", phy_state_name(context->phy_state));
+	LOG_DBG("[%p] phy_state=%s", context->base,
```
Yes but `%p` is for `context->base`.

armandciejak replied:
I changed it in 99237a1472faf2daf552a768ce18daf904a06f3f

armandciejak commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -243,7 +244,8 @@ static void eth_mcux_phy_start(struct eth_context *context)
 {
 #ifdef CONFIG_ETH_MCUX_PHY_EXTRA_DEBUG
-	LOG_DBG("phy_state=%s", phy_state_name(context->phy_state));
+	LOG_DBG("[%p] phy_state=%s", context->base,
```
Yes but `%p` is for `context->base`.

agansari replied:
agree

armandciejak replied:
I changed it in 99237a1472faf2daf552a768ce18daf904a06f3f

armandciejak commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -279,7 +281,8 @@ static void eth_mcux_phy_event(struct eth_context *context)
 
 #ifdef CONFIG_ETH_MCUX_PHY_EXTRA_DEBUG
-	LOG_DBG("phy_state=%s", phy_state_name(context->phy_state));
+	LOG_DBG("[%p] phy_state=%s", context->base,
```
Yes but `%p` is for `context->base`.

armandciejak replied:
I changed it in 99237a1472faf2daf552a768ce18daf904a06f3f

armandciejak committed:
drivers: eth: mcux: Print interface name instead of base register address

With Ethernet controller base address is not straight forward
to figure out if it is ENET or ENET2l, using name is much better.

Signed-off-by: Armand Ciejak <armandciejak@users.noreply.github.com>

agansari reviewed (approved):
Looks good, I need to test this.

armandciejak committed:
drivers: eth: mcux: Fix wrong base address for second controller

Copy paste error, base address of second controller was set to the
same as the first one, corrected to ENET2.

Signed-off-by: Armand Ciejak <armandciejak@users.noreply.github.com>

armandciejak committed:
drivers: eth: mcux: Print interface name in link status messages

Add interface name to link up/down information message
to help tracking which port is changing status.

Signed-off-by: Armand Ciejak <armandciejak@users.noreply.github.com>

pfalcon commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -754,16 +776,17 @@ static void eth_callback(ENET_Type *base, enet_handle_t *handle,
-static void generate_mac(u8_t *mac_addr)
+#if defined(CONFIG_ETH_MCUX_0_RANDOM_MAC) || \
+    defined(CONFIG_ETH_MCUX_1_RANDOM_MAC)
+static void generate_random_mac(u8_t *mac_addr)
```
@bwasim:

> This "generate_random_mac" address function is not limited to the NXP Ethernet driver and can be used by any driver. Therefore I feel ...

While this is a good idea, it's outside the scope of this PR. (Definitely feel free to start working on it in a separate ticket/PR).

pfalcon commented:
@agansari:

> Looks good, I need to test this.

I think it's good idea to approve PR only when it was actually tested. (As an interim measure to give feedback to a PR author that you're ok with their review changes, you can just dismiss -1 review). Otherwise, we risk not properly tested changes to be merged by mistake.

pfalcon reviewed (commented):
@agansari 
> I need to test this in FRDM-K64F (single interface Kinetis).

For reference, I tested master 8bf05c7d61b445f25d722d8e383dcb40a82c9447 + this PR on frdm_k64f, using my usual tests of dumb_http_server & big_http_download samples. Everything works as expected.

Not giving +1, as this PR targets i.MXRT1060, and I can't test that.

armandciejak commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -86,9 +102,11 @@ struct eth_context {
 	enum eth_mcux_phy_state phy_state;
 	bool enabled;
 	bool link_up;
+	u32_t phy_addr;
```
@bwasim OK but phy_addr is a 32-bit unsigned integer parameter in all functions (e.g. ENET_StartSMIWrite).

armandciejak commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -1097,6 +1145,47 @@ static void eth_0_config_func(void)
+	.config_func = eth_1_config_func,
+	.phy_addr = 0U,
+	.phy_duplex = kPHY_FullDuplex,
+	.phy_speed = kPHY_Speed100M,
```
@bwasim It sounds good to me but I think it is out of scope for this PR since I didn't change that part.

bwasim replied:
I understand the concept of keeping changes limited but this is something that should be done eventually..

armandciejak commented on drivers/ethernet/Kconfig.mcux:
```diff
@@ -74,9 +76,49 @@ config ETH_MCUX_0_RANDOM_MAC
+
+if ETH_MCUX_1
+
+choice ETH_MCUX_1_MAC_SELECT
```
@bwasim I didn't change the way the MAC address is handled by the driver, therefore I think your proposal should be treated in a separate PR.

bwasim replied:
If Zephyr project has a issue management system, then we can probably trigger an issue there..

agansari commented:
> I think it's good idea to approve PR only when it was actually tested. (As an interim measure to give feedback to a PR author that you're ok with their review changes, you can just dismiss -1 review). Otherwise, we risk not properly tested changes to be merged by mistake.

You are right @pfalcon , the reason I gave it +1 without testing this time as it's low risk for other platforms, instancing works the same on all platforms.

> For reference, I tested master [8bf05c7](https://github.com/zephyrproject-rtos/zephyr/commit/8bf05c7d61b445f25d722d8e383dcb40a82c9447) + this PR on frdm_k64f, using my usual tests of dumb_http_server & big_http_download samples. Everything works as expected.

Thank you for testing!
